server/src/main/java/org/elasticsearch/gateway/MetaDataStateFormat.java

```diff
@@ -91,86 +92,162 @@ private static void deleteFileIfExists(Path stateLocation, Directory directory,
+                            // in order to write the footer we need to prevent closing the actual index input.
+                        }
+                    })) {
+
+                        builder.startObject();
+                        {
+                            toXContent(builder, state);
+                        }
+                        builder.endObject();
                     }
-                    builder.endObject();
+                    CodecUtil.writeFooter(out);
                 }
-                CodecUtil.writeFooter(out);
-            }
 
-            stateDir.sync(Collections.singleton(tmpFileName));
-            stateDir.rename(tmpFileName, fileName);
-            stateDir.syncMetaData();
-            logger.trace("written state to {}", stateLocation.resolve(fileName));
-        } finally {
-            deleteFileIfExists(stateLocation, stateDir, tmpFileName);
+                stateDir.sync(Collections.singleton(tmpFileName));
+            } catch (Exception e) {
+                // perform clean up only in case of exception, we need to keep directory open and temporary file on disk
+                // if everything is ok for the next algorithm steps
+                performDirectoryCleanup(stateLocation, stateDir, tmpFileName);
+                throw e;
+            }
+            return stateDir;
+        } catch (Exception e) {
+            throw new WriteStateException(false, "failed to write state to the first location tmp file", e);
         }
     }
 
-    private void copyStateToExtraLocation(Directory srcStateDir, Path extraStateLocation, String fileName, String tmpFileName)
-            throws IOException {
-        try (Directory extraStateDir = newDirectory(extraStateLocation)) {
+    private Directory copyStateToExtraLocation(Directory srcStateDir, Path extraStateLocation, String tmpFileName)
+            throws WriteStateException {
+        try {
+            Directory extraStateDir = newDirectory(extraStateLocation);
             try {
                 deleteFileIfExists(extraStateLocation, extraStateDir, tmpFileName);
-                extraStateDir.copyFrom(srcStateDir, fileName, tmpFileName, IOContext.DEFAULT);
+                extraStateDir.copyFrom(srcStateDir, tmpFileName, tmpFileName, IOContext.DEFAULT);
                 extraStateDir.sync(Collections.singleton(tmpFileName));
-                extraStateDir.rename(tmpFileName, fileName);
-                extraStateDir.syncMetaData();
-                logger.trace("copied state to {}", extraStateLocation.resolve(fileName));
-            } finally {
-                deleteFileIfExists(extraStateLocation, extraStateDir, tmpFileName);
+            } catch (Exception e) {
+                // perform clean up only in case of exception, we need to keep directory open and temporary file on disk
+                // if everything is ok for the next algorithm steps
+                performDirectoryCleanup(extraStateLocation, extraStateDir, tmpFileName);
```

Comment:
Perhaps I'm missing something, but it looks like we do `performDirectoryCleanup` on every directory in a `finally` block in the top-level `write` method. Is this extra call also necessary?

Reply:
Same as writeStateToFirstLocation